[JSC] Crash on JSC::Wasm::wasmToJS(unsigned long, unsigned int)

Constellation opened the pull request:
#### 96d7b009c8c18bdea3cc58e0bb3e3a1cb640af39
<pre>
[JSC] Crash on JSC::Wasm::wasmToJS(unsigned long, unsigned int)
<a href="https://bugs.webkit.org/show_bug.cgi?id=285205">https://bugs.webkit.org/show_bug.cgi?id=285205</a>
<a href="https://rdar.apple.com/142111335">rdar://142111335</a>

Reviewed by Justin Michaud.

We should validate functions / exceptions signature when parsing types.

* JSTests/wasm/stress/validate-function-signature-in-section-parser.js: Added.
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseImport):
(JSC::Wasm::SectionParser::parseFunction):

Canonical link: <a href="https://commits.webkit.org/288718@main">https://commits.webkit.org/288718@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fde323968102f7c223599df827d860f4000a368

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84187 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3808 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38491 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89263 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35196 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3896 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11780 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65483 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23319 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87233 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2922 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76504 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45776 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2870 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30734 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34245 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77151 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31498 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90644 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83204 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11454 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8312 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73936 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11680 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72329 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73140 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17449 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2810 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13029 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11406 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16882 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/105623 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11255 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25509 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14731 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13028 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on JSTests/wasm/stress/validate-function-signature-in-section-parser.js:
```diff
@@ -0,0 +1,23 @@
+    };
+    [BigInt64Array,48];
+    v28--;
+    const v90 = new Uint8Array([0,97,115,109,1,0,0,0,1,21,5,96,1,127,0,96,1,127,0,96,1,127,0,96,1,127,0,v28,1,127,0,2,39,5,3,102,111,111,1,48,0,0,1,48,3,102,111,111,0,1,1,48,1,48,0,2,2,49,56,2,45,51,0,3,2,45,51,2,49,56,0,4]);
```
This test should be more simpler? For example, I would like the following:

```js
shouldThrow(function () {
    const bin = new Uint8Array([0,97,115,109,1,0,0,0,1,21,5,96,1,127,0,96,1,127,0,96,1,127,0,96,1,127,0,95,1,127,0,2,39,5,3,102,111,111,1,48,0,0,1,48,3,102,111,111,0,1,1,48,1,48,0,2,2,49,56,2,45,51,0,3,2,45,51,2,49,56,0,4]);
    new WebAssembly.Module(bin);
}, WebAssembly.CompileError);
```

Constellation replied:
Done.

justinmichaud reviewed (approved):
r=me

Constellation committed:
[JSC] Crash on JSC::Wasm::wasmToJS(unsigned long, unsigned int)
https://bugs.webkit.org/show_bug.cgi?id=285205
rdar://142111335

Reviewed by Justin Michaud.

We should validate functions / exceptions signature when parsing types.

* JSTests/wasm/stress/validate-function-signature-in-section-parser.js: Added.
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):
(JSC::Wasm::FunctionParser<Context>::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseImport):
(JSC::Wasm::SectionParser::parseFunction):

Canonical link: https://commits.webkit.org/288718@main
